File: 2024/6/ans.py
from pathlib import Path

# example = """
# ....#.....
# .........#
# ..........
# ..#.......
# .......#..
# ..........
# .#..^.....
# ........#.
# #.........
# ......#...
# """
# input = [list(line) for line in example.split()]

input = []
with open(Path(__file__).with_name('input.txt'), 'r') as file:
    for line in file:
        input.append(list(line[:-1]))

EMPTY = '.'
BLOCKED = '#'
START = '^'

# up, right, down, left
DIRS = ((0, -1), (1, 0), (0, 1), (-1, 0))

empty_set = set()
blocked_set = set()
start_pos = (None, None)


def get_spot(pos):
    x, y = pos
    return input[y][x]


for y in range(len(input)):
    for x in range(len(input[y])):
        pos = (x, y)
        spot = get_spot(pos)
        if spot == EMPTY:
            empty_set.add(pos)
        elif spot == BLOCKED:
            blocked_set.add(pos)
        elif spot == START:
            start_pos = pos
        else:
            raise Exception('unknown char')


reachable_set = set()
current_dir_idx = 0
current_pos = start_pos


def is_in_bounds(pos):
    x, y = pos
    return (0 <= x < len(input)) and (0 <= y < len(input[0]))


def move(pos, dir_idx):
    dir = DIRS[dir_idx]
    # print(f"moved {dir}")
    return (pos[0] + dir[0], pos[1] + dir[1])


def rotate(dir_idx):
    return (dir_idx + 1) % 4


while is_in_bounds(current_pos):
    reachable_set.add(current_pos)
    next_pos = move(current_pos, current_dir_idx)
    while is_in_bounds(next_pos) and get_spot(next_pos) == BLOCKED:
        current_dir_idx = rotate(current_dir_idx)
        next_pos = move(current_pos, current_dir_idx)
    current_pos = next_pos

print(len(reachable_set))

# part 2


def detect_cycle(extra_block: tuple[int, int]) -> bool:
    seen = set()
    current_pos = start_pos
    current_dir = 0

    def step(pos, dir_idx):
        next = move(pos, dir_idx)
        while is_in_bounds(next) and (get_spot(next) == BLOCKED or next == extra_block):
            dir_idx = rotate(dir_idx)
            next = move(pos, dir_idx)
        return next, dir_idx

    while is_in_bounds(current_pos):
        if (current_pos, current_dir) in seen:
            return True
        seen.add((current_pos, current_dir))
        current_pos, current_dir = step(current_pos, current_dir)

    return False


cycle_positions = set()
trial_num = 1  # just for logging
for pos in empty_set:
    print(f"trying extra block position {trial_num}/{len(empty_set)}")
    trial_num += 1
    if detect_cycle(pos):
        cycle_positions.add(pos)

print(len(cycle_positions))

File: 2024/6/ans.go
package main

import (
	"fmt"
	"os"
	"strings"
	"sync"
)

func getInput() []string {
	input, _ := os.ReadFile("input.txt")
    lines := strings.Split(strings.TrimSpace(string(input)), "\n")
    return lines
}

type position = [2]int
type positions = map[position]interface{}

func addPosition(pos position, set *positions) *positions {
	(*set)[pos] = struct{}{}
	return set
}

func getEmptyBlockedAndStart(input *[]string) (empty, blocked positions, start position) {
	empty = positions{}
	blocked = positions{}
	start = position{-1, -1}
	for y, line := range *input {
		for x, spot := range line {
		pos := position{x, y}
		switch spot {
				case '.':
					addPosition(pos, &empty)
				case '#':
					addPosition(pos, &blocked)
				case '^':
					start = pos
				}
		}
	}
	return empty, blocked, start
} 

func getSpot(pos position, input *[]string) rune {
	return rune((*input)[pos[1]][pos[0]])
}

var vectors = [4][2]int{{0, -1}, {1, 0}, {0, 1}, {-1, 0}}
type direction = int


func rotate(dir direction) direction {
	return (dir + 1) % 4
}

func move(pos position, dir direction) position {
	vector := vectors[dir]
	return [2]int{pos[0] + vector[0], pos[1] + vector[1]}
}

func isInBounds(pos position, input *[]string) bool {
	return 0 <= pos[0] && pos[0] < len(*input) && 0 <= pos[1] && pos[1] < len((*input)[0])
}

func isBlocked(pos position, input *[]string) bool {
	return getSpot(pos, input) == '#'
}

func step(pos position, dir direction, input *[]string) (nextpos position, nextdir direction) {
	nextpos = move(pos, dir)
	nextdir = dir
	if isInBounds(nextpos, input) && isBlocked(nextpos, input) {
		nextdir = rotate(dir)
		nextpos = move(pos, nextdir)
	}
	return nextpos, nextdir
}

func getReachable(input *[]string) (reachable positions) {
	reachable = positions{}
	_, _, start := getEmptyBlockedAndStart(input)
	pos := start
	var dir direction = 0 // init to north
	for isInBounds(pos, input) {
		addPosition(pos, &reachable)
		pos, dir = step(pos, dir, input)
	}
	return
}

// part 2 funcs

func stepWithBlock(pos position, dir direction, input *[]string, block position) (nextpos position, nextdir direction) {
	nextpos = move(pos, dir)
	nextdir = dir
	if isInBounds(nextpos, input) && (isBlocked(nextpos, input) || block == nextpos )  {
		nextdir = rotate(dir)
		nextpos = move(pos, nextdir)
	}
	return nextpos, nextdir
}

type state = struct{
	pos position 
	dir direction
}

type states = map[state]interface{}

func hasCycle(input *[]string, block position) bool {
	_, _, start := getEmptyBlockedAndStart(input)
	visited := states{}

	pos := start
	dir := 0

	for isInBounds(pos, input) {
		state := state{pos, dir}
		if visited[state] != nil {
			return true
		}
		visited[state] = struct{}{}
		pos, dir = stepWithBlock(pos, dir, input, block)
	}
	return false
}


func main() {
	input := getInput()

	// part 1
	reachable := getReachable(&input)
	fmt.Println(len(reachable))

	// part 2i
	empty, _, _ := getEmptyBlockedAndStart(&input)
	jobs := make(chan position, len(empty))
	var wg sync.WaitGroup
	var cycles sync.Map
	const workers = 10

	for i := 0; i < workers; i++ {
		wg.Add(1) 
		go func() {
			defer wg.Done() 
			for pos := range jobs { 
				if hasCycle(&input, pos) {
					cycles.Store(pos, struct{}{}) 
				}
			}
		}()
	}

	go func() {
		for pos := range empty {
			jobs <- pos 
		}
		close(jobs) 
	}()

	wg.Wait()
	count := 0
	cycles.Range(func(_, _ interface{}) bool {
		count++
		return true
	})
	fmt.Println(count)
}